src/app/games/hive/HiveRules.ts

```diff
@@ -0,0 +1,237 @@
+                }
+            }
+        }
+        if (state.turn !== 0 && hasNeighbor === false) {
+            return MGPFallible.failure(HiveFailure.MUST_BE_CONNECTED_TO_HIVE());
+        }
+
+        // Pieces must be dropped on an empty space (even the beetle)
+        if (state.getAt(move.coord).isEmpty() === false) {
+            return MGPFallible.failure(HiveFailure.MUST_DROP_ON_EMPTY_SPACE());
+        }
+
+        return MGPFallible.success(undefined);
+    }
+
+    public getPossibleDropLocations(state: HiveState): MGPSet<Coord> {
+        const player: Player = state.getCurrentPlayer();
+        // At turn 0 and 1, the possible drop locations are already known
+        if (state.turn === 0) {
+            return new MGPSet([new Coord(0, 0)]);
+        }
+        if (state.turn === 1) {
+            return new MGPSet(HexagonalUtils.neighbors(new Coord(0, 0)));
+        }
+
+        const remainingPieceOpt: MGPOptional<HivePiece> = state.remainingPieces.getAnyRemainingPiece(player);
+        if (remainingPieceOpt.isAbsent()) {
+            return new MGPSet();
+        }
+
+        const remainingPiece: HivePiece = remainingPieceOpt.get();
+        const locations: MGPSet<Coord> = new MGPSet();
+        // We can only drop next to one of our piece
+        for (const coord of state.occupiedSpaces()) {
+            if (state.getAt(coord).topPiece().owner === player) {
+                for (const neighbor of state.emptyNeighbors(coord)) {
+                    const move: HiveMoveDrop = HiveMoveDrop.from(remainingPiece, neighbor.x, neighbor.y).get();
+                    if (this.isLegalDrop(move, state).isSuccess()) {
+                        locations.add(neighbor);
+                    }
+                }
+            }
+        }
+        return locations;
+    }
+
+    public getPossibleMovesFrom(state: HiveState, coord: Coord): MGPSet<HiveMoveCoordToCoord> {
+        const player: Player = state.getCurrentPlayer();
+        const moves: MGPSet<HiveMoveCoordToCoord> = new MGPSet();
+        const topPiece: HivePiece = state.getAt(coord).topPiece();
+        if (topPiece.owner === player) {
+            for (const move of HivePieceBehaviour.from(topPiece).getPossibleMoves(coord, state)) {
+                if (this.isLegalMoveCoordToCoord(move, state).isSuccess()) {
+                    moves.add(move);
+                }
+            }
+        }
+        return moves;
+    }
+
+    public getPossibleMovesOnBoard(state: HiveState): MGPSet<HiveMoveCoordToCoord> {
+        const moves: MGPSet<HiveMoveCoordToCoord> = new MGPSet();
+        for (const coord of state.occupiedSpaces()) {
+            moves.union(this.getPossibleMovesFrom(state, coord));
+        }
+        return moves;
+    }
+
+    public shouldPass(state: HiveState): boolean {
+        return this.getPossibleDropLocations(state).size() === 0 &&
+            this.getPossibleMovesOnBoard(state).size() === 0;
```

Comment:
aligner horizontalement les this :p ?

Reply:
up !

Reply:
c'est aligné en une seule ligne maintenant :p